flang/lib/Lower/IntrinsicCall.cpp

```diff
@@ -1209,45 +1218,56 @@ mlir::Value toValue(const fir::ExtendedValue &val,
 // IntrinsicLibrary
 //===----------------------------------------------------------------------===//
 
+/// Emit a TODO error message for as yet unimplemented intrinsics.
+static void crashOnMissingIntrinsic(mlir::Location loc, llvm::StringRef name) {
+  mlir::emitError(loc,
+                  "TODO: missing intrinsic lowering: " + llvm::Twine(name));
+  exit(1);
+}
+
 template <typename GeneratorType>
 fir::ExtendedValue IntrinsicLibrary::genElementalCall(
     GeneratorType generator, llvm::StringRef name, mlir::Type resultType,
     llvm::ArrayRef<fir::ExtendedValue> args, bool outline) {
   llvm::SmallVector<mlir::Value> scalarArgs;
-  for (const auto &arg : args) {
-    if (arg.getUnboxed() || arg.getCharBox()) {
+  for (const auto &arg : args)
+    if (arg.getUnboxed() || arg.getCharBox())
       scalarArgs.emplace_back(fir::getBase(arg));
-    } else {
-      // TODO: get the result shape and create the loop...
-      mlir::emitError(loc, "array or descriptor not yet handled in elemental "
-                           "intrinsic lowering");
-      exit(1);
-    }
-  }
+    else
+      mlir::emitError(loc, "nonscalar intrinsic argument");
   if (outline)
     return outlineInWrapper(generator, name, resultType, scalarArgs);
   return invokeGenerator(generator, resultType, scalarArgs);
 }
 
-/// Some ExtendedGenerator operating on characters are also elemental
-/// (e.g LEN_TRIM).
 template <>
 fir::ExtendedValue
 IntrinsicLibrary::genElementalCall<IntrinsicLibrary::ExtendedGenerator>(
     ExtendedGenerator generator, llvm::StringRef name, mlir::Type resultType,
     llvm::ArrayRef<fir::ExtendedValue> args, bool outline) {
   for (const auto &arg : args)
-    if (!arg.getUnboxed() && !arg.getCharBox()) {
-      // TODO: get the result shape and create the loop...
-      mlir::emitError(loc, "array or descriptor not yet handled in elemental "
-                           "intrinsic lowering");
-      exit(1);
-    }
+    if (!arg.getUnboxed() && !arg.getCharBox())
+      mlir::emitError(loc, "nonscalar intrinsic argument");
```

Comment:
I would make this one fir::emitFatalError, no need to continue and risk hitting undefined behaviors later.